libs/pilight/protocols/433.92/quigg_gt9000.c

```diff
@@ -0,0 +1,394 @@
+			createOne(x, x+1);
+		}
+	}
+}
+
+static void initAllCodes(int systemcode, int allcodes[16]) {
+	int i = 0;
+	int systemcode1enc = 0, systemcode2enc = 0, systemcode3enc = 0, systemcode4enc = 0, systemcode5enc = 0;
+	int systemcode1dec = 0, systemcode3dec = 0, systemcode4dec = 0, systemcode5dec = 0;
+	int hash[16] = { 0x0, 0x9, 0xF, 0x4, 0xA, 0xD, 0x5, 0xB,
+			 0x3, 0x2, 0x1, 0x7, 0xE, 0x6, 0xC, 0x8 };
+	
+	systemcode1dec = (systemcode >> 16) & 0xF;
+	//systemcode2dec is always 0, therefore it is not needed
+	//systemcode2dec = (systemcode >> 12) & 0xF;
+	systemcode3dec = (systemcode >> 8) & 0xF;
+	systemcode4dec = (systemcode >> 4) & 0xF;
+	systemcode5dec = systemcode & 0xF;
+	
+	//first 4 bits are not encrypted
+	systemcode1enc = systemcode1dec;
+	
+	//encrypt systemcode
+	for(i=0;i<16;i++) {
+		systemcode2enc = i;
+		systemcode3enc = hash[systemcode2enc]^systemcode3dec;
+		systemcode4enc = hash[systemcode3enc]^systemcode4dec;
+		systemcode5enc = hash[systemcode4enc]^systemcode5dec;
+		allcodes[systemcode5enc] = (systemcode1enc<<16) + (systemcode2enc<<12) + (systemcode3enc<<8) + (systemcode4enc<<4) + systemcode5enc;
+	}
+}
+
+static int createCode(JsonNode *code) {
+	char on1[4] = {10,6,1,5};
+	char on2[4] = {0,3,14,15};
+	char off1[4] = {8,2,11,7};
+	char off2[4] = {4,9,12,13};
+	double itmp = -1;
+	int unit = -1, systemcode = -1, verifysyscode = -1, state = -1, all = 0, statecode = -1;
+	int allcodes[16], binary[RAW_LENGTH/2];
+
+	if(json_find_number(code, "id", &itmp) == 0)
+		systemcode = (int)round(itmp);
+	if(json_find_number(code, "unit", &itmp) == 0)
+		unit = (int)round(itmp);
+	if(json_find_number(code, "all", &itmp) == 0)
+		all = (int)round(itmp);
+	if(json_find_number(code, "off", &itmp) == 0)
+		state=0;
+	else if(json_find_number(code, "on", &itmp) == 0)
+		state=1;
+
+	if((systemcode == -1) || (unit == -1 && all == 0)) {
+		logprintf(LOG_ERR, "quigg_gt9000: insufficient number of arguments");
+		return EXIT_FAILURE;
+	} else if(systemcode == -1 || systemcode < 650000) {
+		logprintf(LOG_ERR, "quigg_gt9000: invalid id range (recommended >650000)");
+		return EXIT_FAILURE;
+	} else if((unit > 4 || unit < 0) && all == 0) {
+		logprintf(LOG_ERR, "quigg_gt9000: invalid unit code range");
+		return EXIT_FAILURE;
+	} else {
+		if(unit == -1 && all == 1) {
+			unit = 5;
+		}
+		quigg_gt9000->rawlen = RAW_LENGTH;
+		//create all 16 codes used by the remote
+		initAllCodes(systemcode, allcodes);
+		//select one of 4 possible codes randomly
+		srand(time(NULL));
+		if(unit == 1 || unit == 3) {
+			if((state == 1 && unit == 1) || (state == 0 && unit == 3))
+				statecode = on1[rand() % 4];
+			else
+				statecode = off1[rand() % 4];
+		} else if(unit == 2 || unit == 4) {
+			if((state == 1 && unit == 2) || (state == 0 && unit == 4))
+				statecode = on2[rand() % 4];
+			else
+				statecode = off2[rand() % 4];
+		} else if(unit == 5) {
+			if(state == 1)
+				statecode = off1[rand() % 4];
+			else
+				statecode = on1[rand() % 4];
```

Comment:
As long as pilight is running yes, create a linked list (check the weather station protocols for it) storing the values of a specific unit/id combination.